### ISSUE TITLE
Feature: Multi-region list file entries

### ISSUE THREAD

**Issue by michihdeu** [TravelMapping/DataProcessing#58]
From Jim: http://tm.teresco.org/forum/index.php?topic=1868.msg7195#msg7195

The idea is that for connected routes that span multiple regions, you could enter something like 

**MA I-90 2 OH I-90 1**

and you would be credited with all I-90 mileage between Exit 2 in Massachusetts to Exit 1 on Ohio, including all of I-90 in NY and PA in between.

**Comment by michihdeu** [comment]
The feature can only work if the wp labels are unique.
I think we should add a data check and eliminate all errors first before introducing the feature.

**Comment by jteresco** [comment]
In this case, there would be a unique region and waypoint label for each endpoint.   As I'm thinking of it, that's enough to figure out which was meant for each endpoint.  How are alt labels relevant?

**Comment by jteresco** [comment]
For my own benefit and by way of explanation, the way I envision this working is that when a line is encountered that has 5 fields instead of 4, like

I-90 MA 2 OH 1 (actual format yet to be determined)

the program would look for a 2 label in the MA and a 1 label in the OH I-90 segments.  It would then find MA and OH in the connected I-90 route and put the needed "normal" segments into the in-memory representation of the user's list for further processing.

The more I think about it, the more I think this isn't too hard.

**Comment by michihdeu** [comment]
If a route crosses the border twice, e.g. B1 from DEU-BB to DEU-BE and back to DEU-BB, there are two segments in BB. Both B1 segments intersect the A10 and I could use "A10" for both wp labels now. If we introduce this feature, it would not work anymore...

**Comment by michihdeu** [comment]
Today, the list file entries are as follows:
DEU-BB B1Pot A10**(22)** BB/BE
DEU-BE B1 BB/BE BE/BB
DEU BB B1 BE/BB A10**(4)**

Again, I could remove (22) and (4) without any problem now...... (of course, I won't change it to "A10" only)

**Comment by jteresco** [comment]
In that case, we could require both region and highway in both parts more like the entry you had in the original comment of this issue.

**Comment by michihdeu** [comment]
Full info for both points: Region, route and wp

`DEU-BB B1Pot A10    DEU-BB B1 A10`

If the end point traveled is in Berlin, it would be, e.g.:

`DEU-BB B1Pot A10    DEU-BE B1 B2`


What would be the rule, that both routes must be in the same line of the _con.csv?
How could a normal user see that it's possible to use the feature?*

If it's not limited one could enter:

`DEU-BW A8 41    DEU-BY A8 DEU/AUT`

A8 from Karlsruhe to Salzburg is interrupted in Munich....

This must be considered here: https://github.com/TravelMapping/DataProcessing/issues/55.

_*I think about a "one click" HB feature to expand the loaded route to all connected routes..._

**Comment by jteresco** [comment]
Some recent forum posts got me thinking about this potential feature again.  Another option could be to keep actual .list files as is, but to add functionality to the web front end to be able to see an entire connected route, click a start and and end travel point, and have it generate a set of .list entries, ready to copy and paste, that would mark all travels between the two points.

**Comment by michihdeu** [comment]
> `DEU-BE B1 B2 DEU-BB B1 A10`

Today's syntax is:
region name (1) + blank + route name + wp name (1) + blank + wp name (2)

Future syntax might be:

region name (1) + blank + route name + wp name (1) + blank + _optional region name (2) + optional blank + optional route name (2) + blank +_ wp name (2)

**Comment by yakra** [comment]
Confused with what you mean with the "blank". I presume it's just just the ordinary space used to delimit the fields (region, route, wp wp), but I'm thrown off a bit by it not being mentioned between `route name (1) + wp name (1)`. Anyway. Assuming that was a mistake...

> region name (1) + blank + route name (1) + wp name (1) + blank + optional region name (2) + optional blank + optional route name (2) + blank + wp name (2)

Yup, this matches the `DEU-BE B1 B2 DEU-BB B1 A10` quoted above, and the `MA I-90 2 OH I-90 1` in the OP.

I had this big long oversized response to this thread mostly typed up on another PC, but TLDR, I think @michihdeu's suggestion here & in the OP would be the way to go for single-line multi-region entries.

**Comment by michihdeu** [comment]
> Assuming that was a mistake...

I've replaced blank by space and fixed the mistake. In addition, the parameters are not optional but only the combination (marked with <>) of parameters is optional.

That means, the relevant string consists of
- 4 blocks (old syntax) or
- 6 blocks (new syntax)

separated by ' 's or TABs and without comment '#'

**Comment by yakra** [comment]
One of the spaces would be optional too, but anyway ;)

> separated by ' 's or TABs and without comment '#'

[TABs are still not implemented](https://github.com/TravelMapping/DataProcessing/issues/213) :(
Easy enough to do, though.

**Comment by michihdeu** [comment]
I wasn't aware that TAB is not valid.
I've also moved the `>` so that the space is included. Space stands for 'separation identifier' whatever it is

**Comment by yakra** [comment]
* [ ] Refactor:
  https://github.com/TravelMapping/DataProcessing/blob/8bcc7644c5dd32821ce38fa042ef69c6d8700266/siteupdate/python-teresco/siteupdate.py#L1315-L1323
  https://github.com/TravelMapping/DataProcessing/blob/f50764d0ee137d66ff32a543c00ff2260d6bf0c0/siteupdate/cplusplus/classes/TravelerList/TravelerList.cpp#L178-L188
into new `Route::store_traveled_segments` function

**Comment by yakra** [comment]
The stage is set for this to be implemented. No more changes to existing code should be required; just gotta write the new stuff & paste it in.

**Comment by yakra** [comment]
## C++ progress
starting off with https://github.com/TravelMapping/UserData/blob/a25a084cc7c12b5e0c42b3e003818caebd3d6257/list_files/michih.list

`diff michih.list michih2.list`
```diff
495a496,540
> DEU-SH A1 5 DEU-NW A1 L115
> DEU-RP A1 117 DEU-SL A1 150
> DEU-NW A2 1 DEU-BB A2 81
> DEU-NW A3 NLD/DEU DEU-BY A3 DEU/AUT
> DEU-HE A4 31 DEU-SN A4 DEU/POL
> DEU-HE A5 1 DEU-BW A5 DEU/CHE
> DEU-SL A6 FRA/DEU DEU-BY A6 DEU/CZE
> DEU-SH A7 DNK/DEU DEU-BY A7Kem DEU/AUT
> DEU-SL A8 LUX/DEU DEU-RP A8 36
> DEU-BW A8 41 DEU-BY A8Aug 82
> DEU-BB A9 1 DEU-BY A9 76
> DEU-BB A10 BE/BB DEU-BE A10 BE/BB
> DEU-MV A11 POL/DEU DEU-BB A11 17
> DEU-BB A13 1 DEU-SN A13 23
> DEU-ST A14 1 DEU-SN A14 38
> DEU-MV A19 4 DEU-BB A19 21
> DEU-SH A20 K7 DEU-BB A20 39
> DEU-SH A23 2 DEU-HH A23 22
> DEU-HH A24 1 DEU-BB A24 26
> DEU-HH A25 1 DEU-SH A25 7
> DEU-NI A27Cux 1 DEU-NI A27 29
> DEU-NI A30Sch NLD/DEU DEU-NW A30Oey 35
> DEU-NI A31 1 DEU-NW A31 42
> DEU-NI A33 5 DEU-NW A33 31
> DEU-NI A36 1 DEU-ST A36 32
> DEU-NI A38 1 DEU-SN A38 33
> DEU-NW A44 51 DEU-HE A44 70
> DEU-NW A45 2 DEU-HE A45Sel 49
> DEU-RP A60Mai 12 DEU-HE A60 30
> DEU-NW A61 16 DEU-BW A61 65
> DEU-SL A62 1 DEU-RP A62 16
> DEU-ST A71 1 DEU-BY A71 31
> DEU-BY A72 1 DEU-SN A72 28
> DEU-TH A73 1 DEU-BY A73 VonTannStr
> DEU-BY A81 1 DEU-BW A81 43
> DEU-BY A96Lin AUT/DEU DEU-BY A96 39
> DEU-BB A111 1 DEU-BE A111 A100
> DEU-BE A113 1 DEU-BB A113 10
> DEU-BB A114 1 DEU-BE A114 4
> DEU-BE A115 1 DEU-BB A115 9
> DEU-BE A117 1 DEU-BB A117 3
> DEU-HH A261 1 DEU-NI A261 4
> DEU-NW A565 1 DEU-RP A565 13
> DEU-HE A643 1 DEU-RP A643 6
> DEU-BW A659 1 DEU-HE A659 HE/BW
498,499d542
< DEU-BW A3 BY/BW BW/BY
< DEU-BW A5 HE/BW DEU/CHE
501,511d543
< DEU-BW A6Man RP/BW BW/HE
< DEU-BW A6 HE/BW BW/BY
< DEU-BW A7 BY/BW BW/BY
< DEU-BW A7Det BY/BW BW/BY
< DEU-BW A7Unt BY/BW BW/BY
< DEU-BW A7Ege BY/BW BW/BY
< DEU-BW A8 41 BW/BY
< DEU-BW A8Lan BY/BW BW/BY
< DEU-BW A61 RP/BW 65
< DEU-BW A81 BY/BW 43
< DEU-BW A96 BY/BW BW/BY
517d548
< DEU-BW A659 1 BW/HE
1049,1058d1079
< DEU-BY A3Asc HE/BY BY/BW
< DEU-BY A3 BW/BY DEU/AUT
< DEU-BY A6 BW/BY DEU/CZE
< DEU-BY A7 HE/BY BY/BW
< DEU-BY A7Neu BW/BY BY/BW
< DEU-BY A7Fel BW/BY BY/BW
< DEU-BY A7Hei BW/BY BY/BW
< DEU-BY A7Kem BW/BY DEU/AUT
< DEU-BY A8Neu BW/BY BY/BW
< DEU-BY A8Aug BW/BY 82
1060,1061d1080
< DEU-BY A9 TH/BY 76
< DEU-BY A45 HE/BY BY/HE
1063,1065d1081
< DEU-BY A71 TH/BY 31
< DEU-BY A72 1 BY/SN
< DEU-BY A73 TH/BY VonTannStr
1067d1082
< DEU-BY A81 1 BY/BW
1074,1075d1088
< DEU-BY A96Lin AUT/DEU BY/BW
< DEU-BY A96 BW/BY 39
1469d1481
< DEU-BE A10 BB/BE BE/BB
1472,1476d1483
< DEU-BE A111 BB/BE A100
< DEU-BE A113 1 BE/BB
< DEU-BE A114 BB/BE 4
< DEU-BE A115 1 BE/BB
< DEU-BE A117 1 BE/BB
1494,1497d1500
< DEU-BB A2 ST/BB 81
< DEU-BB A9 1 BB/ST
< DEU-BB A10 BE/BB BB/BE
< DEU-BB A11 MV/BB 17
1499d1501
< DEU-BB A13 1 BB/SN
1502,1511d1503
< DEU-BB A19 MV/BB 21
< DEU-BB A20Uck MV/BB BB/MV
< DEU-BB A20 MV/BB 39
< DEU-BB A24Kru MV/BB BB/MV
< DEU-BB A24 MV/BB 26
< DEU-BB A111 1 BB/BE
< DEU-BB A113 BE/BB 10
< DEU-BB A114 1 BB/BE
< DEU-BB A115 BE/BB 9
< DEU-BB A117 BE/BB 3
1595,1601d1586
< DEU-HB A1 NI/HB HB/NI
< DEU-HB A1Ars NI/HB HB/NI
< DEU-HB A27But NI/HB HB/NI
< DEU-HB A27Leh NI/HB HB/NI
< DEU-HB A27Bhv NI/HB HB/NI
< DEU-HB A27 NI/HB HB/NI
< DEU-HB A27Mah NI/HB HB/NI
1616,1622d1600
< DEU-HH A1 SH/HH HH/NI
< DEU-HH A7 SH/HH HH/NI
< DEU-HH A7SiW NI/HH HH/NI
< DEU-HH A7SiS NI/HH HH/NI
< DEU-HH A23 SH/HH 22
< DEU-HH A24 1 HH/SH
< DEU-HH A25 1 HH/SH
1624d1601
< DEU-HH A261 1 HH/NI
1637,1646d1613
< DEU-HE A3 RP/HE HE/BY
< DEU-HE A4 31 HE/TH
< DEU-HE A4Her TH/HE HE/TH
< DEU-HE A5 1 HE/BW
< DEU-HE A6 BW/HE HE/BW
< DEU-HE A7Wit NI/HE HE/NI
< DEU-HE A7 NI/HE HE/BY
< DEU-HE A38 NI/HE HE/NI
< DEU-HE A44Die NW/HE HE/NW
< DEU-HE A44 NW/HE 70
1648,1649d1614
< DEU-HE A45 NW/HE HE/BY
< DEU-HE A45Sel BY/HE 49
1651d1615
< DEU-HE A60 RP/HE 30
1658d1621
< DEU-HE A643 1 HE/RP
1660d1622
< DEU-HE A659 BW/HE HE/BW
1886d1847
< DEU-MV A11 POL/DEU MV/BB
1888,1892d1848
< DEU-MV A19 4 MV/BB
< DEU-MV A20 SH/MV MV/BB
< DEU-MV A20Pas BB/MV MV/BB
< DEU-MV A24 SH/MV MV/BB
< DEU-MV A24Suc BB/MV MV/BB
1917,1925d1872
< DEU-NI A1Buc HH/NI NI/HB
< DEU-NI A1Dre HB/NI NI/HB
< DEU-NI A1 HB/NI NI/NW
< DEU-NI A1Osn NW/NI NI/NW
< DEU-NI A2 NW/NI NI/ST
< DEU-NI A7BeW HH/NI NI/HH
< DEU-NI A7BeE HH/NI NI/HH
< DEU-NI A7 HH/NI NI/HE
< DEU-NI A7Han HE/NI NI/HE
1928,1933d1874
< DEU-NI A27Cux 1 NI/HB
< DEU-NI A27SpN HB/NI NI/HB
< DEU-NI A27SpW HB/NI NI/HB
< DEU-NI A27Sch HB/NI NI/HB
< DEU-NI A27Oyt HB/NI NI/HB
< DEU-NI A27 HB/NI 29
1936,1939d1876
< DEU-NI A30Sch NLD/DEU NI/NW
< DEU-NI A30 NW/NI NI/NW
< DEU-NI A31 1 NI/NW
< DEU-NI A33 5 NI/NW
1942,1943d1878
< DEU-NI A38 1 NI/HE
< DEU-NI A38Fri HE/NI NI/TH
1946d1880
< DEU-NI A261 HH/NI 4
1954d1887
< DEU-NI A36 1 NI/ST
2034,2037d1966
< DEU-NW A1Wer NI/NW NW/NI
< DEU-NW A1 NI/NW L115
< DEU-NW A2 1 NW/NI
< DEU-NW A3 NLD/DEU NW/RP
2039,2042d1967
< DEU-NW A30 NI/NW NW/NI
< DEU-NW A30Oey NI/NW 35
< DEU-NW A31 NI/NW 42
< DEU-NW A33 NI/NW 31
2049,2051d1973
< DEU-NW A44 51 NW/HE
< DEU-NW A44War HE/NW NW/HE
< DEU-NW A45 2 NW/HE
2064d1985
< DEU-NW A61 16 NW/RP
2078d1998
< DEU-NW A565 1 NW/RP
2223,2226d2142
< DEU-RP A1 117 RP/SL
< DEU-RP A3 NW/RP RP/HE
< DEU-RP A6 SL/RP RP/BW
< DEU-RP A8 SL/RP 36
2229,2233d2144
< DEU-RP A60Mai 12 RP/HE
< DEU-RP A61 NW/RP RP/BW
< DEU-RP A62Ell SL/RP RP/SL
< DEU-RP A62Wei SL/RP RP/SL
< DEU-RP A62 SL/RP 16
2237d2147
< DEU-RP A565 NW/RP 13
2241d2150
< DEU-RP A643 HE/RP 6
2325,2330d2233
< DEU-SL A1 RP/SL 150
< DEU-SL A6 FRA/DEU SL/RP
< DEU-SL A8 LUX/DEU SL/RP
< DEU-SL A62 1 SL/RP
< DEU-SL A62Noh RP/SL SL/RP
< DEU-SL A62Fre RP/SL SL/RP
2361,2369d2263
< DEU-SN A4 TH/SN DEU/POL
< DEU-SN A9Dob ST/SN SN/ST
< DEU-SN A9 ST/SN SN/ST
< DEU-SN A9Weh ST/SN SN/ST
< DEU-SN A9Kle ST/SN SN/ST
< DEU-SN A9Alt ST/SN SN/ST
< DEU-SN A13 BB/SN 23
< DEU-SN A14Rab ST/SN SN/ST
< DEU-SN A14 ST/SN 38
2371,2372d2264
< DEU-SN A38 ST/SN 33
< DEU-SN A72 BY/SN 28
2408,2414d2299
< DEU-ST A2 NI/ST ST/BB
< DEU-ST A9 BB/ST ST/SN
< DEU-ST A9Loh SN/ST ST/SN
< DEU-ST A9Rub SN/ST ST/SN
< DEU-ST A9Mas SN/ST ST/SN
< DEU-ST A9Gun SN/ST ST/SN
< DEU-ST A9Wei SN/ST ST/TH
2416,2420d2300
< DEU-ST A14 1 ST/SN
< DEU-ST A14Gro SN/ST ST/SN
< DEU-ST A36 NI/ST 32
< DEU-ST A38 TH/ST ST/SN
< DEU-ST A71 1 ST/TH
2469,2471d2348
< DEU-SH A1 5 SH/HH
< DEU-SH A7 DNK/DEU SH/HH
< DEU-SH A20 K7 SH/MV
2473,2474d2349
< DEU-SH A23 2 SH/HH
< DEU-SH A24 HH/SH SH/MV
2510,2515d2384
< DEU-TH A4Ger HE/TH TH/HE
< DEU-TH A4 HE/TH TH/SN
< DEU-TH A9 ST/TH TH/BY
< DEU-TH A38 NI/TH TH/ST
< DEU-TH A71 ST/TH TH/BY
< DEU-TH A73 1 TH/BY
```
```sh
yakra@BiggaTomato:~/TravelMapping/UserData/list_files$ echo -en "additions: "; diff michih.list michih2.list | grep '^>' | wc -l
additions: 45
yakra@BiggaTomato:~/TravelMapping/UserData/list_files$ echo -en "deletions: "; diff michih.list michih2.list | grep '^<' | wc -l
deletions: 176
```
176 - 45 = we're down 131 lines.

---
`~/diff _6f/logs/users/michih.log _6f/logs/users/michih2.log`
```diff
1c1
< Log file created at: Mon May 25 23:55:01 2020
---
> Log file created at: Mon May 25 23:54:58 2020
15c15
< Processed 4652 good lines marking 55840 segments traveled.
---
> Processed 4521 good lines marking 55840 segments traveled.
```
4652 - 4521 = we're down 131 lines.
Same number of segments traveled.
Most importantly, no diffs for the remainder of the logs, meaning @michihdeu's stats are identical.

---
```sh
yakra@BiggaTomato:~/TravelMapping/HighwayData/hwy_data/_systems$ echo $(for roots in $(cut -f5 -d';' deua_con.csv | grep ,); do expr $(echo $roots | tr ',' ' ' | wc -w) - 1; done | tr '\n' '+')0 | bc
132
```
This incantation suggests I should be down 132 lines. In all probability I neglected to delete one when editing the file. I can write another shell script to find out *which* one. ;)

---
There's still plenty of testing to do, plus the Python translation.

**Comment by michihdeu** [comment]
👍

**Comment by yakra** [comment]
michih3.list reverses the order of the points from the new lines added in michih2.list, E.G.
```diff
< DEU-SH A1 5 DEU-NW A1 L115
> DEU-NW A1 L115 DEU-SH A1 5
```
...and stuff is broken. Investigation in progress.

**Comment by yakra** [comment]
```sh
for cr in $(tail -n +2 deua_con.csv); do
  route=$(echo $cr | cut -d';' -f2)
  banner=$(echo $cr | cut -d';' -f3)
  groupName=$(echo $cr | cut -d';' -f4)
  roots=$(echo $cr | cut -d';' -f5 | tr ',' ' ')
  numRoots=$(echo $roots | wc -w)
  echo -n "$route$banner "
  if [[ $groupName != '' ]]; then
    echo -n "($groupName) "
  fi
  echo "has $numRoots root(s)"
  echo -n "Display? (y/n) "
  read disp
  if [[ $disp == 'y' ]]; then
    for root in $roots; do
      firefox "http://192.168.2.3/hb/?units=miles&u=michih3&r=$root"
    done
  fi
done
```
Checked out A1 thru A9. I think I have an idea what's going on.

**Comment by yakra** [comment]
Multi-region .list file support is working as expected, in both C++ & Python.
All that's left to do is add region-splitting support to the C++ version.
Taking a little time to refamiliarize myself with how the FRA split was done.

**HighwayData heads:**
[common ancestor](https://github.com/TravelMapping/HighwayData/commit/345b8c07e175c39512c91e4f3691781f789846a4)
[head used](https://github.com/TravelMapping/HighwayData/commit/1d043ba5c2717ee3b451e71320945f7aa4c8e6c4) in generating [splitregion .list files](forum.travelmapping.net/index.php?topic=3278)
[French Split](https://github.com/TravelMapping/HighwayData/commit/301cfb4d8cda3b9553bf7b13ff486ace50875853)

---
Lines such as
`ESP-CL E80 1(BU30) ITA E-80 A7/A12`
would of course not need to change, as there's nothing referencing FRA itself.
.list file processing won't care whether the intermediate chopped routes are
• ESP-PV and FRA, or
• ESP-PV, FRA-NAQ, FRA-OCC, FRA-NAQ, FRA-OCC and FRA-PAC,
as long as the "before" and "after" CSVs are set up properly. 

We only need to do a rewrite if one or both region fields is FRA (or whatever we're splitting).
For *these* lines, the idea is to output a 6-field connected route entry as would be used after the [region split](https://github.com/TravelMapping/HighwayData/commit/301cfb4d8cda3b9553bf7b13ff486ace50875853) is [complete](https://github.com/TravelMapping/HighwayData/pull/3263) and live on site.
**E.G., `FRA E80 59(A8) ITA E80 SP57` -> `FRA-PAC E80 59(A8) ITA E80 SP57`**
These lines would not be usable before the split is complete, as in the "before" stage, frapac.e80 is in `eurer` in a separate connected route from the one in `eure` that still has ita.e80 (and fra.e80) in it.
Sure, it would be possible to have a solution usable at this stage of the game that outputs multiple .list lines, but that goes against the goal of having clean intuitive data for the end user, and would be a more of a PitA to code. So nope.

**Comment by yakra** [comment]
* [ ] **old HighwayData:**
  michihFRA.list & michihFRA6.list should produce identical log files
* [ ] **new HighwayData:**
  michihFRAs.list & michihFRA6s.list should produce identical log files

**Comment by michihdeu** [comment]
> Taking a little time to refamiliarize myself with how the FRA split was done.

Do you plan to have a converter to "clean-up" user list files?

I don't request it but just out of curiosity.....

> Multi-region .list file support is working, in both C++ & Python.

So, I could enter multi-region entries when the pull request will be submitted and merged? If so, what needs to be added to the manual? Is anything required to be done in the web frontend?

**Comment by yakra** [comment]
@michihdeu [wrote](https://github.com/TravelMapping/DataProcessing/issues/58#issuecomment-635208083):
> Do you plan to have a converter to "clean-up" user list files?
>
> I don't request it but just out of curiosity.....

No. As far as .list files go, I'm just extending the functionality that writes new [splitregion .list files](http://forum.travelmapping.net/index.php?topic=3278) to be able to work with the new 6-field line format.

> So, I could enter multi-region entries when the pull request will be submitted and merged?

Yes. For deua, you can use the lines [above](https://github.com/TravelMapping/DataProcessing/issues/58#issuecomment-633801071). I could even commit the changes in michih2.list into michih.list on GitHub, if you're OK with somebody else making edits to your .list file.

> If so, what needs to be added to the manual?

The existing 4-field line format continues to work as it has.
A new 6-field format is now supported:
`Region1 Route1 Waypoint1 Region2 Route2 Waypoint2`
The two routes must be in the same connected route.
I'm not fully up to date on the [LONG LIVE THE MANUAL](forum.travelmapping.net/index.php?topic=3172) thread, but discussion on how to integrate the info & instructions needed into [participate.php](travelmapping.net/devel_new/participate.php) is probably best saved for the forum.

> Is anything required to be done in the web frontend?

***Required***, per se, not as such.
The changes are all on the DataProcessing side; by the time the info makes it into the DB, it's all the same data stored the same way.
Any **helpful features** that can be done in the web frontend, to make it easier for travelers to write their .list lines? Yes.
https://github.com/TravelMapping/Web/issues/421
https://github.com/TravelMapping/Web/issues/192
**Either or both** of these would be helpful. https://github.com/TravelMapping/Web/issues/192 would be quicker & easier to implement.
**Neither** is *required*; this would work as soon as the pull request is merged; no need to wait for hb/web features to make it easier. Those could happen later.

---
That all said, I'm not quite ready to make the pull request just yet.
A little testing remains to be done, and there are a few things I'd like to fine-tune...

---
* [ ] **unambiguous userlog error messages**
  Right now, the .list line
  `NH I-95 44 ME I-95 44`
  would produce the following userlog error message:
  `Waypoint label 44 not found in line: NH I-95 44 ME I-95 44`
  This is ambiguous; which `44` do we mean?
  A clearer solution is to include the whole region/route/label combo:
  `Waypoint NH I-95 44 not found in line: NH I-95 44 ME I-95 44`

---
* **region-splits for 4-field lines**
  I [wrote](https://github.com/TravelMapping/DataProcessing/issues/58#issuecomment-635063532):
  > In theory, I could change the existing code for 4-field chopped route lines so that instead of
  > [multiple 4-field lines]
  > we'd get
  > [a single 6-field line]
  > ...
  > Is it worth the bother?

I think *yes* now.
* It would allow for cleaner code overall.
  * [ ] Currently using the existing splitregion.cpp and a new file splitregion_connected.cpp
    The code in splitregion_connected.cpp could become the new splitregion.cpp, and I could `#include` it for both types of lines with a few `#define`s and `#undef`s.
    This *should* work, right?
  * [ ] `ConnectedRoute::list_lines` and `Route::list_line` would no longer be needed, and could be removed. Though the latter may be useful in the future; perhaps I could leave it in...
* [ ] Continuing to use a 4-field line when the new entry is in only one new chopped region should be a simple modification.
* [ ] In the forum, I [wrote](http://forum.travelmapping.net/index.php?topic=2976.msg13711#msg13711):
  > New lines are listed in "canonical" order from the beginning of the connected route to the end...
  > I *could* enhance this so that the "direction" of the original .listed segment is retained, but I'm thinking this is too much labor for too little gain.
  
  This is now easier with the changes made in #330.

**Comment by michihdeu** [comment]
> > If so, what needs to be added to the manual?
>
> The existing 4-field line format continues to work as it has.
> A new 6-field format is now supported:
> `Region1 Route1 Waypoint1 Region2 Route2 Waypoint2`
> The two routes must be in the same connected route.
> I'm not fully up to date on the LONG LIVE THE MANUAL thread, but discussion on how to integrate the info & instructions needed into participate.php is probably best saved for the forum.

> That all said, I'm not quite ready to make the pull request just yet.
> A little testing remains to be done, and there are a few things I'd like to fine-tune...

Ok, I plan to publish participate.php on Monday.

I'll add the extension for 6-field format later - when you say you are ready. We could have it as a hidden feature for a few days so that some contributors can test it first.

> unambiguous userlog error messages
...
> This is ambiguous; which 44 do we mean?

You could indicate the field - e.g. `in 6th field`
What is output when there are 7 fields? You could also output 7 fields but limited to 6 or whatever.

**Comment by yakra** [comment]
> We could have it as a hidden feature for a few days so that some contributors can test it first.

Good idea.

> > unambiguous userlog error messages
> ...
> > This is ambiguous; which 44 do we mean?
>
> You could indicate the field - e.g. `in 6th field`

Already changed to output the whole region/route/label combo:
`Waypoint NH I-95 44 not found in line: NH I-95 44 ME I-95 44`

> What is output when there are 7 fields?

Counting only the number of fields before the `#` comment delimiter, a line with any number of fields other than 0, 4 or 6 will produce an `Incorrect format line: ` error.

**Comment by michihdeu** [comment]
> other than 0, 4 or 6 will produce an `Incorrect format line:` error.

yep, you could change it to `Incorrect format line (3 fields instead of 4 or 6):` or `Incorrect format line (7 fields instead of 4 or 6):`

**Comment by yakra** [comment]
* [ ] Another good idea.
  I may phrase it
  `Incorrect format line (4 or 6 fields expected, found 7):`

**Comment by michihdeu** [comment]
yep! Much better than my quick draft 👍

**Comment by michihdeu** [comment]
I thought about adding log file error examples to the manual but I omitted it... so far... because it would be redundant to the log message and keeping it up-to-date would be a thing if anything will be changed in the future. When the error log is not clear, we should improve it instead of describing it on the manual....

**Comment by yakra** [comment]
Agreed. Userlog error messages should be clear and self-explanatory.

**Comment by yakra** [comment]
* [x] Convert all .list files from 4-field to 6-field for testing purposes,
  both as-listed and in reverse waypoint order.
  In UserData/list_files/:
```sh
git checkout 83a47aee090102cd598b385d8174fd9ed4513415
F='+\?\*\?[-a-zA-Z0-9()/_\.]\+'
D='[ \t]\+'
users=$(ls *.list | grep -v __ | sed 's~.list$~~')
for user in $users; do
  cat $user.list | tr '\r' '\n' \
  | sed -e 's~[ \t]*#.*~~' \
  | sed -e "s~^\($F\)$D\($F\)$D\($F\)$D\($F\)$~\1 \2 \3 \1 \2 \4~" \
  > "$user"__6.list
  cat $user.list | tr '\r' '\n' \
  | sed -e 's~[ \t]*#.*~~' \
  | sed -e "s~^\($F\)$D\($F\)$D\($F\)$D\($F\)$~\1 \2 \4 \1 \2 \3~" \
  > "$user"__6r.list
done
```

* [ ] Run `siteupdate -p splitregion FRA` with HighwayData @ 1d043ba5c2717ee3b451e71320945f7aa4c8e6c4
  * [ ] **CRASH!** Complete with backtrace & memory map. Happens with `rebelgtp__6.list` and `rebelgtp__6r.list`. This will take some debugging.

* [ ] User logs should have no diffs other than error messages.
  In logs/users/:
```sh
users=$(ls *.log | grep -v __ | sed 's~.log$~~')
for user in $users; do
  offset4=$(grep -n -m 1 'good lines marking' $user.log | cut -f1 -d:)
  offset6=$(grep -n -m 1 'good lines marking' "$user"__6.log | cut -f1 -d:)
  offset6r=$(grep -n -m 1 'good lines marking' "$user"__6r.log | cut -f1 -d:)
  ( echo "$user.log vs "$user"__6.log"
    diff <(tail -n +$offset4 $user.log) <(tail -n +$offset6 "$user"__6.log) \
    | sed -e 's~^\([0-9]\+,\?[0-9]*[acd][0-9]\+,\?[0-9]*\)$~\\e[33m\1\\e[0m~' \
          -e 's~^\(<.*\)~\\e[35m\1\\e[0m~' \
          -e 's~^\(>.*\)~\\\e[34m\1\\e[0m~' \
          -e 's~^---$~\\e[37m---\\e[0m~' \
    | xargs -0 echo -e
    echo "$user.log vs "$user"__6r.log"
    diff <(tail -n +$offset4 $user.log) <(tail -n +$offset6r "$user"__6r.log) \
    | sed -e 's~^\([0-9]\+,\?[0-9]*[acd][0-9]\+,\?[0-9]*\)$~\\e[33m\1\\e[0m~' \
          -e 's~^\(<.*\)~\\e[35m\1\\e[0m~' \
          -e 's~^\(>.*\)~\\\e[34m\1\\e[0m~' \
          -e 's~^---$~\\e[37m---\\e[0m~' \
    | xargs -0 echo -e
  )
done | less -r
```

* [ ] Run `siteupdate` using splitregion .lists with HighwayData @ 1d043ba5c2717ee3b451e71320945f7aa4c8e6c4

* [ ] Run the 2nd shell script again. All we should see are differing numbers of good lines.

**Comment by yakra** [comment]
* [ ] Rethink this:
  ```c++
  HighwaySegment *orig_hs = r1->segment_list[index1?index1-1:0];
  ```
  Just because I don't *see* any bugs doesn't mean there aren't any.

**Comment by yakra** [comment]
Tests are looking good.
Committed to my local repo on BiggaTomato.
Before I push to GitHub, I want to do some investigation and see if I happened to fix the [Schrodinger's user logs diffs](https://github.com/TravelMapping/DataProcessing/issues/323) bug in the process.

**Comment by michihdeu** [comment]
> > So, I could enter multi-region entries when the pull request will be submitted and merged?
>
> Yes. For deua, you can use the lines above. I could even commit the changes in michih2.list into michih.list on GitHub, if you're OK with somebody else making edits to your .list file.

The question was only related to required additional feature implementation. It was answered, there isn't. Multi-region entries will work with the pull request of this feature.

I have no problem with edits to my .list file. I thought about it but since I've sorted my .list file by regions.... Where should I put them? I could do it by country but there are also E routes... I could do it by systems but that's not worth the effort. I don't convert it and keep it as-is. I could use the new feature for new entries but would end up with the same issue where to put it.

The feature is mostly for _normal_ users who complaint about the complexity especially when there are short segments in different regions which appeared when DEU, ESP and FRA were split.

The feature is perfect when you sort your .list file by systems or travels.

**Comment by yakra** [comment]
When I did it, I put the new entries after `# GERMANY`, before `# BADEN-WUERTTEMBERG`.
Each system has an associated country, even if the "country" is `eur`...
If grouping by region, one could group it with the lowest-level region encompassing the connected route, or the clinched part of it, or whatever. ...If the .list file is set up that way, of course, Not everybody's going to organize by region and have a section for EUR. Or DEU. I don't have one for USA or CAN.

Not to sound like I'm suggesting how to organize your .list. I have no business doing that of course; I'm just thinking out loud.

> The feature is perfect when you sort your .list file by systems or travels.

A while back I started keeping my larger trips sorted by travels. LOL, on my last trip I had to consistently cross state lines on state routes (But then any I- or US route I'm likely to use, I've had clinched for years).
If I go back to 2013, there's I-90...
